homeassistant/components/sensoterra/sensor.py

```diff
@@ -0,0 +1,149 @@
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    PERCENTAGE,
+    SIGNAL_STRENGTH_DECIBELS_MILLIWATT,
+    EntityCategory,
+    UnitOfTemperature,
+)
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import CONFIGURATION_URL, DOMAIN
+from .coordinator import SensoterraCoordinator, SensoterraSensor
+from .models import ProbeSensorType
+
+type SensoterraConfigEntry = ConfigEntry[SensoterraCoordinator]
+
+
+SENSORS: tuple[SensorEntityDescription, ...] = (
+    SensorEntityDescription(
+        key=ProbeSensorType.MOISTURE,
+        state_class=SensorStateClass.MEASUREMENT,
+        suggested_display_precision=0,
+        device_class=SensorDeviceClass.MOISTURE,
+        native_unit_of_measurement=PERCENTAGE,
+        translation_key="soil_moisture_at_cm",
+    ),
+    SensorEntityDescription(
+        key=ProbeSensorType.SI,
+        state_class=SensorStateClass.MEASUREMENT,
+        suggested_display_precision=1,
+        name="SI @ {depth} cm",
+    ),
+    SensorEntityDescription(
+        key=ProbeSensorType.TEMPERATURE,
+        state_class=SensorStateClass.MEASUREMENT,
+        suggested_display_precision=0,
+        device_class=SensorDeviceClass.TEMPERATURE,
+        native_unit_of_measurement=UnitOfTemperature.CELSIUS,
+    ),
+    SensorEntityDescription(
+        key=ProbeSensorType.BATTERY,
+        state_class=SensorStateClass.MEASUREMENT,
+        suggested_display_precision=0,
+        device_class=SensorDeviceClass.BATTERY,
+        native_unit_of_measurement=PERCENTAGE,
+        entity_category=EntityCategory.DIAGNOSTIC,
+    ),
+    SensorEntityDescription(
+        key=ProbeSensorType.RSSI,
+        state_class=SensorStateClass.MEASUREMENT,
+        suggested_display_precision=0,
+        device_class=SensorDeviceClass.SIGNAL_STRENGTH,
+        native_unit_of_measurement=SIGNAL_STRENGTH_DECIBELS_MILLIWATT,
+        entity_category=EntityCategory.DIAGNOSTIC,
+    ),
+    SensorEntityDescription(
+        key=ProbeSensorType.LASTSEEN,
+        device_class=SensorDeviceClass.TIMESTAMP,
+        entity_category=EntityCategory.DIAGNOSTIC,
+        translation_key="last_seen",
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: SensoterraConfigEntry,
+    async_add_devices: AddEntitiesCallback,
+) -> None:
+    """Set up Sensoterra sensor."""
+
+    coordinator = entry.runtime_data
+
+    @callback
+    def _async_add_devices(devices: dict[str, SensoterraSensor]) -> None:
+        async_add_devices(
+            SensoterraEntity(coordinator, sensor, description)
+            for sensor in devices.values()
+            for description in SENSORS
+            if sensor.type == description.key
```

Comment:
Well, this is what I implemented initially (https://github.com/home-assistant/core/blob/0ca493510af19733ee7f9ab98eeae0a71d9588cb/homeassistant/components/sensoterra/sensor.py), but @zweckj proposed to change it into a list of tuples because it already contained a `key` with the same value (https://github.com/home-assistant/core/pull/119642#discussion_r1641940031). What I could do is to revert back to a dictionary and construct the `key` attribute from the dictionary key. Something like 
```python
self.entity_description = description
self.entity_description.key = sensor.type
```
Would that be better?

Reply:
No because entity descriptions are frozen. That little duplication of the key isn't the worst imo